### DISCUSSION TITLE
Change the density of fluid in *custum_scene.py

### DISCUSSION THREAD

**Discussion by BoyuanTang331** [InteractiveComputerGraphics/SPlisHSPlasH#273] [Q&A] [answered]
the density of the fluid has already been defined from the custum_scene.py as 1000km/m^3

how could I change this attribute of the fluid, I tried to make a small change as complex_scene.py,  but it shows some bugs "Process finished with exit code -1073741819 (0xC0000005)"

`import pysplishsplash as sph
import pysplishsplash.Utilities.SceneLoaderStructs as Scenes

def main():
    
    base = sph.Exec.SimulatorBase()
    base.init(useGui=True,  sceneFile=sph.Extras.Scenes.Empty)
    sim = sph.Simulation.getCurrent()
    
    gui = sph.GUI.Simulator_GUI_imgui(base)
    base.setGui(gui)
    base.setValueString(base.PARTICLE_EXPORT_ATTRIBUTES, "velocity;density")
   
    scene = sph.Exec.SceneConfiguration.getCurrent().getScene()
    scene.particleRadius = 0.015
    # scene.sim2D = True
    scene.boundaryModels.append(Scenes.BoundaryData(meshFile="../models/UnitBox.obj", translation=[0., 3.0, 0.], scale=[4., 6., 4.], color=[0.1, 0.4, 0.5, 1.0], isWall=True, mapInvert=True, mapResolution=[25, 25, 25]))
    scene.fluidBlocks.append(Scenes.FluidBlock(id='Fluid', boxMin = [-1.5, 0.0, -1.5], boxMax = [-0.5, 1.0, -0.5], mode=0, initialVelocity=[0.0, 0.0, 0.0]))
 

    fluid = sim.getFluidModel(0)
    fluid.setValueFloat(fluid.DENSITY0, 800.0)
    
    base.run()

if __name__ == "__main__":
    main()
`

by the way, is there any simply method to change the density from FluidBlock or FluidModel

**Comment by janbender** [comment]
Instead of ```base.run()``` use ```base.initSimulation()``` to initialize the fluid model and start the simulation with ``` base.runSimulation()```. Finally, call ```base.cleanup()``` to cleanup everything after the simulation. 

If you call ```fluid = sim.getFluidModel(0)``` before the fluid model is initialized you get the error. ```base.run()``` just calls 

``` 
base.initSimulation()
base.runSimulation()
base.cleanup()
```

This is shorter but does not allow to change the parameters of a fluid model. 

Here is a corrected version of your code:

```
import pysplishsplash as sph
import pysplishsplash.Utilities.SceneLoaderStructs as Scenes


def main():
    base = sph.Exec.SimulatorBase()
    base.init(useGui=True,  sceneFile=sph.Extras.Scenes.Empty)
    sim = sph.Simulation.getCurrent()

    gui = sph.GUI.Simulator_GUI_imgui(base)
    base.setGui(gui)
    base.setValueString(base.PARTICLE_EXPORT_ATTRIBUTES, "velocity;density")

    scene = sph.Exec.SceneConfiguration.getCurrent().getScene()
    scene.particleRadius = 0.015
    # scene.sim2D = True
    scene.boundaryModels.append(Scenes.BoundaryData(meshFile="../models/UnitBox.obj", translation=[0., 3.0, 0.], scale=[4., 6., 4.], color=[0.1, 0.4, 0.5, 1.0], isWall=True, mapInvert=True, mapResolution=[25, 25, 25]))
    scene.fluidBlocks.append(Scenes.FluidBlock(id='Fluid', boxMin = [-1.5, 0.0, -1.5], boxMax = [-0.5, 1.0, -0.5], mode=0, initialVelocity=[0.0, 0.0, 0.0]))

    base.initSimulation()

    fluid = sim.getFluidModel(0)
    fluid.setValueFloat(fluid.DENSITY0, 800.0)

    base.runSimulation()
    base.cleanup()


if __name__ == "__main__":
    main()
```

**Reply by BoyuanTang331** [reply]
Hi，I have an additional question about the boundary, in the Scene.boundarymodels it uses directly unitBox.obj,  how to define the thickness of the wall？Should I use scale=[] to zoom the ratio or is there any easy method? And what is the initial thickness of the wall in unibox.obj? I have changed the camera_position and get a view that there is thickness in the surrounding walls but at the bottom, it may not have a thickness  
![image](https://github.com/InteractiveComputerGraphics/SPlisHSPlasH/assets/117408630/b128bf87-18a6-439c-a71d-c921df3aa9a2)


 I have tried to read the information from unitBox.obj, there are three variables  ‘v’，‘vt’，‘f’.  what is the use of vt here in the boundary？Can I define the box boundary without texture (delete the 'vt')？

Best regards

**Comment by janbender** [comment, marked as answer]
Typically the thickness is 0. For particle based boundary handling a wall is just a thin geometry where the particles should not go through. For implicit boundary handling (volume maps or density maps) you can artificially increase the thickness of the wall by the key "mapThickness" in the scene file. See also here: 

https://splishsplash.readthedocs.io/en/latest/file_format.html#rigidbodies
